test/unittest/unittest_nntrainer_tensor_v2_fp16.cpp

```diff
@@ -394,6 +394,963 @@ TEST(nntrainer_Tensor, initialize_08_p) {
+                            nntrainer::Tformat::NCHW,
+                            nntrainer::Tdatatype::FP16);
+
+  GEN_TEST_INPUT(input, i * (batch * height) + j * (width) + k);
+
+  nntrainer::TensorV2 original;
+  original.copy(input);
+
+  status = input.multiply_i(2.0);
+  EXPECT_EQ(status, ML_ERROR_NONE);
+
+  _FP16 *data = original.getData<_FP16>();
+  ASSERT_NE(nullptr, data);
+  _FP16 *indata = input.getData<_FP16>();
+  ASSERT_NE(nullptr, indata);
+
+  for (int i = 0; i < batch * height * width * channel; ++i) {
+    EXPECT_FLOAT_EQ(data[i] + data[i], indata[i]);
+  }
+}
+
+TEST(nntrainer_Tensor, multiply_i_02_fp16_p) {
+  int status = ML_ERROR_NONE;
+  int batch = 3;
+  int channel = 1;
+  int height = 3;
+  int width = 10;
+
+  nntrainer::TensorV2 input(batch, channel, height, width,
+                            nntrainer::Tformat::NCHW,
+                            nntrainer::Tdatatype::FP16);
+  GEN_TEST_INPUT(input, i * (batch * height) + j * (width) + k);
+
+  nntrainer::TensorV2 original;
+  original.copy(input);
+
+  status = input.multiply_i(input);
+  EXPECT_EQ(status, ML_ERROR_NONE);
+
+  _FP16 *data = original.getData<_FP16>();
+  ASSERT_NE(nullptr, data);
+  _FP16 *indata = input.getData<_FP16>();
+  ASSERT_NE(nullptr, indata);
+
+  for (int i = 0; i < batch * height * width * channel; ++i) {
+    EXPECT_FLOAT_EQ(data[i] * data[i], indata[i]);
+  }
+}
+
+TEST(nntrainer_Tensor, multiply_i_03_fp16_n) {
+  int status = ML_ERROR_NONE;
+  int batch = 3;
+  int channel = 1;
+  int height = 3;
+  int width = 10;
+
+  nntrainer::TensorV2 input(batch, channel, height, width,
+                            nntrainer::Tformat::NCHW,
+                            nntrainer::Tdatatype::FP16);
+  GEN_TEST_INPUT(input, i * (batch * height) + j * (width) + k);
+
+  nntrainer::TensorV2 target2(batch, channel, height - 2, width - 1,
+                              nntrainer::Tformat::NCHW,
+                              nntrainer::Tdatatype::FP16);
+  status = input.multiply_i(target2);
+
+  EXPECT_EQ(status, ML_ERROR_INVALID_PARAMETER);
+}
+
+TEST(nntrainer_Tensor, multiply_i_broadcast_01_fp16_p) {
+  {
+    nntrainer::TensorDim ref_dim(3, 2, 4, 5, nntrainer::Tformat::NCHW,
+                                 nntrainer::Tdatatype::FP16);
+    nntrainer::TensorV2 t = rangedV2(3, 2, 4, 5, nntrainer::Tformat::NCHW,
+                                     nntrainer::Tdatatype::FP16);
+    nntrainer::TensorV2 m = rangedV2(1, 2, 4, 5, nntrainer::Tformat::NCHW,
+                                     nntrainer::Tdatatype::FP16);
+    _FP16 answer_data[] = {static_cast<_FP16>(0),    static_cast<_FP16>(1),
+                           static_cast<_FP16>(4),    static_cast<_FP16>(9),
+                           static_cast<_FP16>(16),   static_cast<_FP16>(25),
+                           static_cast<_FP16>(36),   static_cast<_FP16>(49),
+                           static_cast<_FP16>(64),   static_cast<_FP16>(81),
```

Comment:
all these static_cast to _FP16 is making readability very low.
if anyone has a good idea, please give me a suggestion!

Reply:
why don't you make Macro or Function that can convert to FP16?

Reply:
for example 
```cpp
template <typename T> // T models Any
struct static_cast_func
{
  template <typename T1> // T1 models type statically convertible to T
  T operator()(const T1& x) const { return static_cast<T>(x); }
};
```
```cpp
std::transform(char_buff, char_buff + len, float_buff, static_cast_func<float>());
```
like this

Reply:
thanks :) this really helped a lot